react-tic-tac-toe/src/board.tsx

```diff
@@ -0,0 +1,193 @@
+    private handleGameStateChange(newState: GameState) {
+        var event = new CustomEvent("gameStateChange", { "detail": this.state.gameState });
+        event.initEvent("gameStateChange", false, true); 
+        window.dispatchEvent(event);
+    }   
+    
+    // check the game state - use the latest move
+    private checkGameState(cells: CellValue[], latestPos: number, latestVal: CellValue): GameState {
+        if (this.state.gameState !== "") {
+            return this.state.gameState;
+        }
+        
+        // check row
+        let result = this.check3Cells(cells, 3 * Math.floor(latestPos / 3), 
+            3 * Math.floor(latestPos / 3) + 1, 3 * Math.floor(latestPos/3) + 2);
+        if (result) {
+            return result; 
+        }
+        
+        // check col
+        result = this.check3Cells(cells, latestPos % 3, latestPos % 3 + 3, latestPos % 3 + 6);
+        if (result) {
+            return result;
+        }
+        
+        // check diag
+        result = this.check3Cells(cells, 0, 4, 8);
+        if (result) {
+            return result;
+        }
+        result = this.check3Cells(cells, 2, 4, 6);
+        if (result) {
+            return result;
+        }
+        
+        // check draw - if all cells are filled
+        if (this.findAllEmptyCells(cells).length === 0) {
+            return "Draw";          
+        }
+                
+        return "";
+    }
+    
+    // check if 3 cells have same non-empty val - return the winner state; otherwise undefined 
+    private check3Cells(cells: CellValue[], pos0: number, pos1: number, pos2: number): GameState {
+        if (cells[pos0] === cells[pos1] &&
+            cells[pos1] === cells[pos2] &&
+            cells[pos0] !== "") {
+            if (cells[pos0] === "X") {
+                return "X Wins!";
+            }
+            return "O Wins!";
+        }
+        else {
+            return undefined;
+        }
+    }
+    
+    // list all empty cell positions
+    private findAllEmptyCells(cells : CellValue[]): number[] {
+        return cells.map((v, i) => { 
+            if (v === "") {
+                return i;
+            }
+            else { 
+                return undefined;
+            }
+        }).filter(v => { return v !== undefined });        
+    }
+    
+    // make a move
+    private move(pos: number, val: CellValue, callback?: () => void): void {
+        if (this.state.gameState === "" &&
+            this.state.cells[pos] === "") {
+            let newCells = this.state.cells.slice();
+            newCells[pos] = val;
+            let oldState = this.state.gameState;
+            this.setState({cells: newCells, gameState: this.checkGameState(newCells, pos, val)}, () => {
+                if (this.state.gameState !== oldState) {
+                    this.handleGameStateChange(this.state.gameState);
+                }
+                if (callback) {        
+                    callback.call(this);
```

Comment:
Sorry, what I meant by my previous comment is that you don't need to bind `this` to the callback anyhow, so we should get rid of that.